### DISCUSSION TITLE
Remove rangeStrategy=auto, pin

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#15766] [Suggest an Idea]
When Renovate was created - npm only at the time - it was my opinion that dependencies should nearly always be pinned. rangeStrategy=pin and rangeStrategy=auto were part of that.

Many have used this feature, but also many have been confused by it. I think that as use of dependency automation shifts from early adopters to the majority, it's maybe time to remove this "opinionated" concept and instead simply have a config option `pinVersions` which people can opt into using `packageRules`.

This means:
- Remove `rangeStrategy=auto` and logic completely
- Remove `rangeStrategy=pin` but migrate it to `pinVersions=true`

**Comment by rarkins** [comment]
Here's our allowed values today with my annotations:

```
      'auto', <- remove
      'pin', <- migrate
      'bump', <- always update, but bump the range lower bound
      'replace', <- only update if the range needs to change
      'widen', <- widen the existing range if the new version falls outside it
      'update-lockfile', <- always update, even if the range doesn't need to change
      'in-range-only', <- update only if the range doesn't need to change
```

`replace` could be expressed as "ignore in-range updates". `in-range-only` could be expressed "ignore out-of-range updates". These could potentially be `packageRules` options if we expose enough metadata, e.g. `inRangeUpdate` true/false or `rangeUpdates=any/in-range-only/in-range-excluded`

New default behavior:
- rangeStrategy=replace
- rangeUpdates=any

Common changes:
- Set rangeStrategy to bump or widen
- Set rangeUpdates to in-range-excluded to restore old behaviour
- Set rangeUpdates to in-range-only to match rangeStrategy=in-range-only

**Reply by rarkins** [reply]
One complication: I think we might auto detect whether to widen ranges today if they are complex ranges. Need to look into this more.

**Comment by deviantintegral** [comment]
I'll 👍 this, as in our projects we always set `rangeStrategy=bump`. This is because we expect our automation tools not to break local tools, especially for junior developers. Every guide on composer (for example) will assume that `composer update` or `composer update <package>` will get you the latest compatible version, without needing to touch `composer.json`. Sure, that's less useful if a project is keeping updated, but it's fairly common for us to inherit projects that are behind and it can take weeks or even months to remediate things.
